syndeo-template/src/main/java/com/google/cloud/syndeo/transforms/datagenerator/DataGeneratorSchemaTransform.java

```diff
@@ -0,0 +1,103 @@
+ * the License.
+ */
+package com.google.cloud.syndeo.transforms.datagenerator;
+
+import com.google.common.flogger.GoogleLogger;
+import java.io.Serializable;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Parser;
+import org.apache.beam.sdk.schemas.transforms.SchemaTransform;
+import org.apache.beam.sdk.schemas.utils.AvroUtils;
+import org.apache.beam.sdk.transforms.FlatMapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.PeriodicImpulse;
+import org.apache.beam.sdk.transforms.Reshuffle;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionRowTuple;
+import org.apache.beam.sdk.values.Row;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+public class DataGeneratorSchemaTransform implements SchemaTransform, Serializable {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private long recordsPerSecond = 1000l;
+  private long minutesToRun = 10;
+  private String schema;
+
+  public DataGeneratorSchemaTransform(long recordsPerSecond, long minutesToRun, String schema) {
+    this.recordsPerSecond = recordsPerSecond;
+    this.minutesToRun = minutesToRun;
+    this.schema = schema;
+  }
+
+  @Override
+  public PTransform<PCollectionRowTuple, PCollectionRowTuple> buildTransform() {
+    return new PTransform<PCollectionRowTuple, PCollectionRowTuple>() {
+      @Override
+      public PCollectionRowTuple expand(PCollectionRowTuple input) {
+        org.apache.beam.sdk.schemas.Schema beamSchema =
+            AvroUtils.toBeamSchema(new Parser().parse(schema));
+        final Instant startTime = Instant.now();
+        PCollection<Instant> instants =
+            input
+                .getPipeline()
+                .apply(
+                    PeriodicImpulse.create()
+                        .startAt(startTime)
+                        .stopAt(Instant.now().plus(Duration.standardMinutes(minutesToRun)))
+                        .withInterval(Duration.millis(1000))
+                        .applyWindowing());
+        instants = instants.apply(Reshuffle.viaRandomKey());
+
+        PCollection rows =
+            instants
+                .apply(FlatMapElements.via(new InstantToRowFn(startTime, schema)))
+                .setRowSchema(beamSchema);
+        return PCollectionRowTuple.of("output", rows);
+      }
+    };
+  }
+
+  private class InstantToRowFn extends SimpleFunction<Instant, List<Row>> {
+
+    private Instant startTime;
+    private String schema;
+    private String format;
+
+    public InstantToRowFn(Instant startTime, String schema) {
+      this.startTime = startTime;
+      this.schema = schema;
+    }
+
+    @Override
+    public List<Row> apply(Instant input) {
+      final Schema schema = Schema.parse(this.schema);
```

Comment:
might be good to avoid parsing the schema on every record? instead only parsing it in the constructor?

Reply:
Avro Schema is not serializable, which needs to be created when apply is called. Will consider using serializable schema later.